File: src/main/java/com/intetics/lukyanenko/models/OrderDetail.java
package com.intetics.lukyanenko.models;

import java.lang.ref.WeakReference;

public class OrderDetail extends Common
{
  private int                  id;
  private GoodsItem            goodsItem;
  private WeakReference<Order> order;
  private double               quantity;
  private int                  goodsItemId;
  private int                  orderId;
  private double               itemPrice;
  
  public int getId()
  {
    return id;
  }
  
  public void setId(int id)
  {
    this.id = id;
  }
  
  public GoodsItem getGoodsItem()
  {
    return goodsItem;
  }
  
  public void setGoodsItem(GoodsItem goodsItem)
  {
    this.goodsItem = goodsItem;
    this.goodsItemId = goodsItem.getId();
  }
  
  public Order getOrder()
  {
    return order.get();
  }
  
  public void setOrder(Order order)
  {
    this.order = new WeakReference<Order>(order);
    this.orderId = order.getId();
  }
  
  public double getQuantity()
  {
    return quantity;
  }
  
  public void setQuantity(double quantity)
  {
    this.quantity = quantity;
  }
  
  public int getGoodsItemId()
  {
    return goodsItemId;
  }
  
  public void setGoodsItemId(int goodsItemId)
  {
    this.goodsItemId = goodsItemId;
    if (goodsItem != null && goodsItem.getId() != goodsItemId)
      goodsItem = null;
  }
  
  public int getOrderId()
  {
    return orderId;
  }
  
  public void setOrderId(int orderId)
  {
    this.orderId = orderId;
  }
  
  public double getItemPrice()
  {
    if (goodsItem == null)
      return itemPrice;
    else
      return goodsItem.getPrice();
  }
  
  public void setItemPrice(double itemPrice)
  {
    this.itemPrice = itemPrice;
  }
}

File: src/main/java/com/intetics/lukyanenko/models/OrderDetail.kt
package com.intetics.lukyanenko.models

class OrderDetail : Common() {
    var id: Int = 0
    var goodsItem: GoodsItem? = null
    var order: Order? = null
    var quantity: Double = 0.toDouble()
    var itemPrice: Double = 0.toDouble()
        get() = goodsItem?.price ?: field

    var goodsItemId: Int? = null
        get() = goodsItem?.id ?: field
        set(value) {
            if (value != goodsItemId) order = null
            field = value
        }

    var orderId: Int? = null
        get() = order?.id ?: field
        set(value) {
            if (value != orderId) order = null
            field = value
        }
}
